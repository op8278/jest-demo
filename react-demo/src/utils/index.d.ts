import { ShallowWrapper } from 'enzyme'

export function findTestWrapper(wrapper: ShallowWrapper, selector: any): ShallowWrapper
