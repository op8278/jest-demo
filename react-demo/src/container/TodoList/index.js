import React, { Component } from 'react'
import axios from 'axios'

import Header from './components/Header'
import UndoList from './components/UndoList'

import './index.css'

class TodoList extends Component {
  constructor(props) {
    super(props)
    this.state = {
      undoList: [],
    }
  }

  // 1. axios
  // componentDidMount() {
  //   axios
  //     .get('/undoList.json')
  //     .then((res) => {
  //       const undoList = res.data
  //       this.setState({ undoList })
  //     })
  //     .catch(() => {})
  // }

  // 2. axios + setTimeout
  componentDidMount() {
    setTimeout(() => {
      axios
        .get('/undoList.json')
        .then((res) => {
          const undoList = res.data
          this.setState({ undoList })
        })
        .catch(() => {})
    }, 4000)
  }

  addUndoItem = (value) => {
    this.setState({
      undoList: [
        ...this.state.undoList,
        {
          status: 'div',
          value,
        },
      ],
    })
  }

  deleteItem = (index) => {
    const { undoList } = this.state
    undoList.splice(index, 1)
    this.setState({
      undoList,
    })
  }

  changeStatus = (index) => {
    const newList = this.state.undoList.map((item, listIndex) => {
      if (index === listIndex) {
        return {
          ...item,
          status: 'input',
        }
      }
      return {
        ...item,
        status: 'div',
      }
    })
    this.setState({
      undoList: newList,
    })
  }

  handleBlur = (index) => {
    const newList = this.state.undoList.map((item, listIndex) => {
      if (index === listIndex) {
        return {
          ...item,
          status: 'div',
        }
      }
      return item
    })
    this.setState({
      undoList: newList,
    })
  }

  valueChange = (index, value) => {
    const { undoList } = this.state
    undoList[index].value = value
    this.setState({
      undoList,
    })
  }

  render() {
    const { undoList } = this.state

    return (
      <div>
        <Header addUndoItem={this.addUndoItem} />
        <UndoList
          list={undoList}
          deleteItem={this.deleteItem}
          changeStatus={this.changeStatus}
          handleBlur={this.handleBlur}
          valueChange={this.valueChange}
        />
      </div>
    )
  }
}

export default TodoList
