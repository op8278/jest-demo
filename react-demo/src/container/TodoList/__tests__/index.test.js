import React from 'react'
import { shallow, mount } from 'enzyme'

import TodoList from '../index'
import { findTestWrapper } from '../../../utils'

describe('TodoList 单元测试', () => {
  describe('初始化', () => {
    test('快照, undoList 数组 为空', () => {
      const wrapper = mount(<TodoList />)

      wrapper.setState({
        undoList: [],
      })

      expect(wrapper).toMatchSnapshot()
    })

    test('快照, undoList 数组有值,且不含 input', () => {
      const wrapper = mount(<TodoList />)

      wrapper.setState({
        undoList: [
          { status: 'div', value: 'value1' },
          { status: 'div', value: 'value2' },
        ],
      })

      expect(wrapper).toMatchSnapshot()
    })

    test('快照, undoList 数组有值,且含 input', () => {
      const wrapper = mount(<TodoList />)

      wrapper.setState({
        undoList: [
          { status: 'div', value: 'value1' },
          { status: 'input', value: 'value2' },
        ],
      })

      expect(wrapper).toMatchSnapshot()
    })

    test('只包含一个 Header 组件', () => {
      const wrapper = shallow(<TodoList />)
      const Header = wrapper.find('Header')
      expect(Header).toHaveLength(1)
    })
    test('只包含一个 UndoList 组件', () => {
      const wrapper = shallow(<TodoList />)
      const UndoList = wrapper.find('UndoList')
      expect(UndoList).toHaveLength(1)
    })

    test('默认 undoList 为数组为空', () => {
      const wrapper = shallow(<TodoList />)
      expect(wrapper.state('undoList')).toEqual([])
    })
    test('应该给 Header 组件传递 addUndoItem 方法', () => {
      const wrapper = shallow(<TodoList />)
      const Header = wrapper.find('Header')

      // 这种方式耦合性太强,不推荐
      // expect(Header.prop("addUndoItem")).toBe(wrapper.instance().addUndoItem);

      // 解耦
      expect(Header.prop('addUndoItem')).toBeTruthy()
    })

    test('应该给 UndoList 组件传递 list 数据和 deleteItem,changeStatus,handleBlur,valueChange 方法', () => {
      const wrapper = shallow(<TodoList />)
      const UndoList = wrapper.find('UndoList')

      expect(UndoList.prop('deleteItem')).toBeTruthy()
      expect(UndoList.prop('changeStatus')).toBeTruthy()
      expect(UndoList.prop('handleBlur')).toBeTruthy()
      expect(UndoList.prop('valueChange')).toBeTruthy()
      expect(UndoList.prop('list')).toBeTruthy()
    })
  })

  describe('行为', () => {
    test('触发 addUndoItem 函数时,新增 undoList 内容', () => {
      const wrapper = shallow(<TodoList />)

      const value = '新增内容'

      /* 单位测试写法,没有跟其他单位(组件)耦合 */
      const { addUndoItem } = wrapper.instance()

      /* 集成测试写法,耦合度高 */
      // const Header = wrapper.find("Header");
      // const addUndoItem = Header.prop("addUndoItem");

      addUndoItem(value)

      expect(wrapper.state('undoList').length).toBe(1)
      expect(wrapper.state('undoList')[0]).toEqual({
        status: 'div',
        value,
      })
    })

    test('触发 deleteItem 函数时,删除 undoList 内容', () => {
      const wrapper = shallow(<TodoList />)

      wrapper.setState({
        undoList: [
          { status: 'div', value: 'ddd' },
          { status: 'div', value: 'aaa' },
        ],
      })

      expect(wrapper.state('undoList').length).toBe(2)

      const { deleteItem } = wrapper.instance()

      // 删除索引1
      deleteItem(1)

      expect(wrapper.state('undoList').length).toBe(1)
      expect(wrapper.state('undoList')).toEqual([{ status: 'div', value: 'ddd' }])
    })

    test('触发 changeStatus 函数时,undoList 对应内容数据项的 status 应该被修改', () => {
      const wrapper = shallow(<TodoList />)

      wrapper.setState({
        undoList: [
          { status: 'div', value: 'ddd' },
          { status: 'div', value: 'aaa' },
        ],
      })

      const { changeStatus } = wrapper.instance()

      // 改变索引1
      changeStatus(1)

      expect(wrapper.state('undoList')).toEqual([
        { status: 'div', value: 'ddd' },
        { status: 'input', value: 'aaa' },
      ])
    })

    test('触发 handleBlur 函数时,undoList 对应内容数据项的 status 应该从 input 恢复成 div', () => {
      const wrapper = shallow(<TodoList />)

      wrapper.setState({
        undoList: [
          { status: 'div', value: 'ddd' },
          { status: 'input', value: 'aaa' },
        ],
      })

      const { handleBlur } = wrapper.instance()

      // 改变索引1
      handleBlur(1)

      expect(wrapper.state('undoList')).toEqual([
        { status: 'div', value: 'ddd' },
        { status: 'div', value: 'aaa' },
      ])
    })

    test('触发 valueChange 函数时,undoList 对应内容数据项的 value 应该被修改', () => {
      const wrapper = shallow(<TodoList />)

      wrapper.setState({
        undoList: [
          { status: 'div', value: 'ddd' },
          { status: 'div', value: 'aaa' },
        ],
      })

      const { valueChange } = wrapper.instance()

      // 改变索引1
      valueChange(1, '我是修改内容')

      expect(wrapper.state('undoList')).toEqual([
        { status: 'div', value: 'ddd' },
        { status: 'div', value: '我是修改内容' },
      ])
    })
  })
})
