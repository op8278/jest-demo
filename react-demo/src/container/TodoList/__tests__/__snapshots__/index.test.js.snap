// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList 单元测试 初始化 快照, undoList 数组 为空 1`] = `
<TodoList>
  <div>
    <Header
      addUndoItem={[Function]}
    >
      <div
        className="header"
      >
        <div
          className="header-content"
        >
          TodoList
          <input
            className="header-input"
            data-test="input"
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="Todo"
            value=""
          />
        </div>
      </div>
    </Header>
    <UndoList
      changeStatus={[Function]}
      deleteItem={[Function]}
      handleBlur={[Function]}
      list={Array []}
      valueChange={[Function]}
    >
      <div
        className="undo-list"
      >
        <div
          className="undo-list-title"
        >
          正在进行
          <div
            className="undo-list-count"
            data-test="count"
          >
            0
          </div>
        </div>
        <ul
          className="undo-list-content"
        />
      </div>
    </UndoList>
  </div>
</TodoList>
`;

exports[`TodoList 单元测试 初始化 快照, undoList 数组有值,且不含 input 1`] = `
<TodoList>
  <div>
    <Header
      addUndoItem={[Function]}
    >
      <div
        className="header"
      >
        <div
          className="header-content"
        >
          TodoList
          <input
            className="header-input"
            data-test="input"
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="Todo"
            value=""
          />
        </div>
      </div>
    </Header>
    <UndoList
      changeStatus={[Function]}
      deleteItem={[Function]}
      handleBlur={[Function]}
      list={
        Array [
          Object {
            "status": "div",
            "value": "value1",
          },
          Object {
            "status": "div",
            "value": "value2",
          },
        ]
      }
      valueChange={[Function]}
    >
      <div
        className="undo-list"
      >
        <div
          className="undo-list-title"
        >
          正在进行
          <div
            className="undo-list-count"
            data-test="count"
          >
            2
          </div>
        </div>
        <ul
          className="undo-list-content"
        >
          <li
            className="undo-list-item"
            data-test="list-item"
            key="0"
            onClick={[Function]}
          >
            value1
            <div
              className="undo-list-delete"
              data-test="delete-item"
              onClick={[Function]}
            >
              -
            </div>
          </li>
          <li
            className="undo-list-item"
            data-test="list-item"
            key="1"
            onClick={[Function]}
          >
            value2
            <div
              className="undo-list-delete"
              data-test="delete-item"
              onClick={[Function]}
            >
              -
            </div>
          </li>
        </ul>
      </div>
    </UndoList>
  </div>
</TodoList>
`;

exports[`TodoList 单元测试 初始化 快照, undoList 数组有值,且含 input 1`] = `
<TodoList>
  <div>
    <Header
      addUndoItem={[Function]}
    >
      <div
        className="header"
      >
        <div
          className="header-content"
        >
          TodoList
          <input
            className="header-input"
            data-test="input"
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="Todo"
            value=""
          />
        </div>
      </div>
    </Header>
    <UndoList
      changeStatus={[Function]}
      deleteItem={[Function]}
      handleBlur={[Function]}
      list={
        Array [
          Object {
            "status": "div",
            "value": "value1",
          },
          Object {
            "status": "input",
            "value": "value2",
          },
        ]
      }
      valueChange={[Function]}
    >
      <div
        className="undo-list"
      >
        <div
          className="undo-list-title"
        >
          正在进行
          <div
            className="undo-list-count"
            data-test="count"
          >
            2
          </div>
        </div>
        <ul
          className="undo-list-content"
        >
          <li
            className="undo-list-item"
            data-test="list-item"
            key="0"
            onClick={[Function]}
          >
            value1
            <div
              className="undo-list-delete"
              data-test="delete-item"
              onClick={[Function]}
            >
              -
            </div>
          </li>
          <li
            className="undo-list-item"
            data-test="list-item"
            key="1"
            onClick={[Function]}
          >
            <input
              autoFocus={true}
              data-test="input"
              onBlur={[Function]}
              onChange={[Function]}
              value="value2"
            />
            <div
              className="undo-list-delete"
              data-test="delete-item"
              onClick={[Function]}
            >
              -
            </div>
          </li>
        </ul>
      </div>
    </UndoList>
  </div>
</TodoList>
`;
